From d782634a84e73d1a4e002c52d89194d44cdc98d2 Mon Sep 17 00:00:00 2001
From: Vedant Paranjape <22630228+VedantParanjape@users.noreply.github.com>
Date: Thu, 14 Oct 2021 11:14:22 +0530
Subject: [PATCH 5/8] Implement watchdog style menu timeout

Standard GRUB implementation cannot resist spurious keypress and stops booting.
This change resets timeout to its initial value. Grub menutimeout
should be set to some comfort value e.g. 5 sec

Signed-off-by: Michael Malyshev <mikem@zededa.com>
Signed-off-by: Vedant Paranjape <22630228+VedantParanjape@users.noreply.github.com>
---
 grub-core/normal/menu.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/grub-core/normal/menu.c b/grub-core/normal/menu.c
index 8397886fa..c1a5d7c06 100644
--- a/grub-core/normal/menu.c
+++ b/grub-core/normal/menu.c
@@ -579,6 +579,7 @@ run_menu (grub_menu_t menu, int nested, int *auto_boot)
   int default_entry, current_entry;
   int timeout;
   enum timeout_style timeout_style;
+  int initial_timeout_value;
 
   default_entry = get_entry_number (menu, "default");
 
@@ -659,6 +660,7 @@ run_menu (grub_menu_t menu, int nested, int *auto_boot)
 
   current_entry = default_entry;
 
+  initial_timeout_value = grub_menu_get_timeout ();
  refresh:
   menu_init (current_entry, menu, nested);
 
@@ -702,9 +704,9 @@ run_menu (grub_menu_t menu, int nested, int *auto_boot)
 	{
 	  if (timeout >= 0)
 	    {
-	      grub_env_unset ("timeout");
-	      grub_env_unset ("fallback");
-	      clear_timeout ();
+	      /* reset timeout value if the user press any key */
+	      grub_menu_set_timeout(initial_timeout_value);
+	      menu_print_timeout (initial_timeout_value);
 	    }
 
 	  switch (c)
-- 
2.25.1

